src/wp-includes/class-wp-theme-json-resolver.php

```diff
@@ -90,6 +99,25 @@ protected static function read_json_file( $file_path ) {
 		return $config;
 	}
 
+	/**
+	 * Generates a new cache key to determine when to clear theme cache.
+	 *
+	 * @since 6.1.0
+	 *
+	 * @return string A cache key.
+	 */
+	protected static function get_new_theme_cache_key() {
+		$registry = WP_Block_Type_Registry::get_instance();
+		$blocks   = $registry->get_all_registered();
+
+		/*
+		 * Generate the key based on the current number of blocks registered.
+		 * This ensures that Theme JSON data accessed at registration time
+		 * does not result in stale block data.
+		 */
+		return 'registered-blocks-' . count( $blocks );
```

Comment:
Can this result in a case where we end up with the same number of blocks but they're different? Perhaps we could get a hash of the `$blocks` array instead.

Reply:
Yeah, that seems to be a concern shared by some folks: https://github.com/WordPress/wordpress-develop/pull/3373#issuecomment-1263806141

Reply:
Unfortunately we can't rely on the registered block names because there are instances where we unregister an old version of a block and re-register a new one with the same name.

https://github.com/WordPress/wordpress-develop/pull/3373#issuecomment-1263888253

Reply:
> Unfortunately we can't rely on only the registered block names because there is at least one instance where we unregister an old version of a block and re-register a new one with the same name within the same function.

Thank you for pointing that out! I just tried to address that at https://github.com/WordPress/wordpress-develop/pull/3359#discussion_r986762129.